### PR TITLE
DAOS-10830 test: Add recovery/ddb.py to test ddb

### PR DESCRIPTION
**Pull request by shimizukko**
Test the following ddb subcommands:
ls: test_ls()
rm: test_rm()
load: test_load()
dump_value: test_dump_value()

See code or the ticket for the test steps.

Add container_list_objects() in daos_utils.py

Skip-unit-tests: true
Skip-fault-injection-test: true
Test-tag: ddb
Signed-off-by: Makito Kano <makito.kano@intel.com>

### CODE REVIEW

File: src/tests/ftest/SConscript [hunk 1]
```diff
@@ -10,7 +10,7 @@ def scons():
 
     env.Install(ftest_install_dir, Glob('*.*'))
 
-    dirs = ['aggregation', 'fault_injection', 'checksum',
+    dirs = ['aggregation', 'fault_injection', 'cat_recovery', 'checksum',
```
Comment (daltonbohning) [inline src/tests/ftest/SConscript:13]: How do you feel about simplifying the new directory name to `recovery`?

File: src/tests/ftest/util/daos_utils.py [hunk 2]
```diff
@@ -604,6 +604,25 @@ def container_list_attrs(self, pool, cont, sys_name=None, verbose=False):
             ("container", "list-attrs"), pool=pool, cont=cont, sys_name=sys_name,
             verbose=verbose)
 
+    def container_list_objects(self, pool, cont, sys_name=None):
+        """Call daos container list-objects.
+
+        Args:
+            pool (str): Pool UUID.
+            cont (str): Container UUID.
```
Comment (daltonbohning) [inline src/tests/ftest/util/daos_utils.py:612]: Suggestion - commands accept UUID or labels
```suggestion
            pool (str): Pool UUID or label
            cont (str): Container UUID or label
```

File: src/tests/ftest/cat_recovery/ddb.py [hunk 3]
```diff
@@ -0,0 +1,666 @@
...
+from pydaos.raw import IORequest, DaosObjClass
+from general_utils import create_string_buffer, report_errors, run_pcmd, \
+    distribute_files, DaosTestError, get_clush_command, run_command
+from ddb_utils import DdbCommand
+from exception_utils import CommandFailure
+
+
+SAMPLE_DKEY = "Sample dkey"
+SAMPLE_AKEY = "Sample akey"
+SAMPLE_DATA = "Sample data"
```
Comment (daltonbohning) [inline src/tests/ftest/cat_recovery/ddb.py:21]: Can we put these in the config? Or maybe generate random strings?

File: src/tests/ftest/cat_recovery/ddb.py [hunk 4]
```diff
@@ -0,0 +1,666 @@
...
+        super().__init__(*args, **kwargs)
+        self.ioreqs = []
+        self.dkeys = []
+        self.akeys = []
+        self.data_list = []
+
+    def insert_objects(self, object_count, dkey_count, akey_count):
+        """Insert objects, dkeys, akeys, and data into the container.
+
+        Inserted objects: self.ioreqs
```
Comment (daltonbohning) [inline src/tests/ftest/cat_recovery/ddb.py:41]: We should fix this spelling error. Maybe by adding an entry to words.dict

File: src/tests/ftest/cat_recovery/ddb.py [hunk 5]
```diff
@@ -0,0 +1,666 @@
...
+
+    def __init__(self, *args, **kwargs):
+        """Initialize a DdbTest object."""
+        super().__init__(*args, **kwargs)
+        self.ioreqs = []
+        self.dkeys = []
+        self.akeys = []
+        self.data_list = []
+
+    def insert_objects(self, object_count, dkey_count, akey_count):
```
Comment (daltonbohning) [inline src/tests/ftest/cat_recovery/ddb.py:38]: Could this function be genericized and moved to a util file?

File: src/tests/ftest/cat_recovery/ddb.py [hunk 6]
```diff
@@ -0,0 +1,666 @@
...
+                vos_file = file
+                break
+
+        if not vos_file:
+            self.fail(
+                "vos file wasn't found in /mnt/daos/{}".format(self.pool.uuid.lower()))
+        else:
+            self.log.info("vos_file: %s", vos_file)
+
+        return vos_file
```
Comment (daltonbohning) [inline src/tests/ftest/cat_recovery/ddb.py:112]: Recommend simplifying
```suggestion
        for file in cmd_out[0]["stdout"]:
            # Assume the VOS file has "vos" in the file name.
            if "vos" in file:
                self.log.info("vos_file: %s", file)
                return file

        self.fail(
            "vos file wasn't found in /mnt/daos/{}".format(self.pool.uuid.lower()))
            
        return None  # to appease pylint
```

File: src/tests/ftest/cat_recovery/ddb.py [hunk 7]
```diff
@@ -0,0 +1,666 @@
...
+
+    def copy_remote_to_local(self, remote_file_path):
```
Comment (daltonbohning) [inline src/tests/ftest/cat_recovery/ddb.py:114]: Also wonder if this could be moved to a util file

File: src/tests/ftest/cat_recovery/ddb.py [hunk 8]
```diff
@@ -0,0 +1,666 @@
...
+        current_file_path = "".join([remote_file_path, ".", self.hostlist_servers[0]])
+        mv_command = "mv {} {}".format(current_file_path, remote_file_path)
+        try:
+            run_command(command=mv_command)
+        except DaosTestError as error:
+            raise CommandFailure(
+                "ERROR: Moving {} to {}: {}".format(
+                    current_file_path, remote_file_path, error)) from error
+
+    def test_ls(self):
```
Comment (daltonbohning) [inline src/tests/ftest/cat_recovery/ddb.py:142]: Recommend making the test name more specific / less prone to existing elsewhere.
Similar for other test methods in this file
```suggestion
    def test_recovery_ddb_ls(self):
```

File: src/tests/ftest/cat_recovery/ddb.py [hunk 9]
```diff
@@ -0,0 +1,666 @@
...
+        self.get_dmg_command().system_stop()
+
+        # 1. Verify container UUID.
+        cmd_result = ddb_command.list_component()
+        # Sample output.
+        # [0] d4e0c836-17bd-4df3-b255-929732486bab
+        # stdout is a list which contains each line as separate element. Concatenate them
+        # to single string so that we can apply regex.
+        ls_out = "\n".join(cmd_result[0]["stdout"])
+        match = re.findall(r"(\[\d+\])\s+([a-f0-9\-]+)", ls_out)
```
Comment (daltonbohning) [inline src/tests/ftest/cat_recovery/ddb.py:189]: You should be able to just do this
```suggestion
        match = re.findall(r"(\[\d+\])\s+([a-f0-9\-]+)", cmd_result[0]["stdout"], re.MULTILINE)
```

File: src/tests/ftest/cat_recovery/ddb.py [hunk 10]
```diff
@@ -0,0 +1,666 @@
...
+        self.ioreqs[0].obj.close()
+        self.container.close()
+        self.pool.disconnect()
+        self.pool.connect()
+        self.container.open()
+        self.ioreqs[0].obj.open()
+
+        # 8. Verify the data in the akey with single_fetch().
+        data_size = len(new_data)
+        data_size += 1
```
Comment (daltonbohning) [inline src/tests/ftest/cat_recovery/ddb.py:562]:
```suggestion
        data_size = len(new_data) + 1
```

**Review by daltonbohning — COMMENTED** [review]
- Recommend fixing the spelling errors if you need to repush

File: src/tests/ftest/SConscript (hunk 1 above)
Comment (shimizukko) [reply src/tests/ftest/SConscript]: Done

File: src/tests/ftest/util/daos_utils.py (hunk 2 above)
Comment (shimizukko) [reply src/tests/ftest/util/daos_utils.py]: Done

File: src/tests/ftest/cat_recovery/ddb.py (hunk 3 above)
Comment (shimizukko) [reply src/tests/ftest/cat_recovery/ddb.py]: I updated the code to generate random string for these.

File: src/tests/ftest/cat_recovery/ddb.py (hunk 4 above)
Comment (shimizukko) [reply src/tests/ftest/cat_recovery/ddb.py]: Done

File: src/tests/ftest/cat_recovery/ddb.py (hunk 5 above)
Comment (shimizukko) [reply src/tests/ftest/cat_recovery/ddb.py]: Done

File: src/tests/ftest/cat_recovery/ddb.py (hunk 6 above)
Comment (shimizukko) [reply src/tests/ftest/cat_recovery/ddb.py]: Done. Thanks

File: src/tests/ftest/cat_recovery/ddb.py (hunk 7 above)
Comment (shimizukko) [reply src/tests/ftest/cat_recovery/ddb.py]: Done

File: src/tests/ftest/cat_recovery/ddb.py (hunk 8 above)
Comment (shimizukko) [reply src/tests/ftest/cat_recovery/ddb.py]: Done

File: src/tests/ftest/cat_recovery/ddb.py [hunk 11]
```diff
@@ -0,0 +1,666 @@
...
+        4. Verify there is one akey for every dkey. Also verify the key string and the
+        size.
+        5. Verify data length in every akey.
+        6. Restart the server for the cleanup.
+        7. Reset the container and the pool to prepare for the cleanup.
+
+        :avocado: tags=all,weekly_regression
+        :avocado: tags=vm
+        :avocado: tags=cat_rec
+        :avocado: tags=ddb,ddb_ls
```
[the remark this answers is not in the document]
Comment (shimizukko) [reply src/tests/ftest/cat_recovery/ddb.py]: Done

File: src/tests/ftest/cat_recovery/ddb.py (hunk 9 above)
Comment (shimizukko) [reply src/tests/ftest/cat_recovery/ddb.py]: `cmd_result[0]["stdout"]` is a list that contains each line as separate element, so this approach didn't work.

File: src/tests/ftest/cat_recovery/ddb.py [hunk 12]
```diff
@@ -0,0 +1,666 @@
...
+        cmd_result = ddb_command.list_component()
+        # Sample output.
+        # [0] d4e0c836-17bd-4df3-b255-929732486bab
+        # stdout is a list which contains each line as separate element. Concatenate them
+        # to single string so that we can apply regex.
+        ls_out = "\n".join(cmd_result[0]["stdout"])
+        match = re.findall(r"(\[\d+\])\s+([a-f0-9\-]+)", ls_out)
+        self.log.info("List container match = %s", match)
+
+        actual_uuid = match[0][1].lower()
```
[the remark this answers is not in the document]
Comment (shimizukko) [reply src/tests/ftest/cat_recovery/ddb.py]: Done

File: src/tests/ftest/cat_recovery/ddb.py (hunk 10 above)
Comment (shimizukko) [reply src/tests/ftest/cat_recovery/ddb.py]: Done

File: src/tests/ftest/util/general_utils.py [hunk 13]
```diff
@@ -1525,3 +1525,94 @@ def set_avocado_config_value(section, key, value):
...
+    """
+    # Use clush --rcopy to copy the file from the remote server node to the local test
+    # node. clush will append .<server_hostname> to the file when copying.
+    args = "--rcopy {} --dest {}".format(remote_file_path, test_dir)
+    clush_command = get_clush_command(hosts=remote, args=args)
+    try:
+        run_command(command=clush_command)
+    except DaosTestError as error:
+        print("ERROR: Copying {} from {}: {}".format(remote_file_path, remote, error))
+        raise error
```
Comment (daltonbohning) [inline src/tests/ftest/util/general_utils.py:1607]: I don't know if we should be using `print` here.
@phender Might have a suggestion.

File: src/tests/ftest/recovery/ddb.py [hunk 14]
```diff
@@ -170,11 +84,13 @@ def test_ls(self):
...
-        dkey_count = self.params.get("dkey_count", "/run/*")
-        akey_count = self.params.get("akey_count", "/run/*")
-        self.insert_objects(
-            object_count=object_count, dkey_count=dkey_count, akey_count=akey_count)
+        object_count = 5
+        dkey_count = 2
+        akey_count = 1
+        insert_objects(
+            context=self.context, container=self.container, object_count=5,
+            dkey_count=2, akey_count=1, base_dkey=self.random_dkey,
```
Comment (daltonbohning) [inline src/tests/ftest/recovery/ddb.py:92]: Why not pass the variables?
```suggestion
        object_count = 5
        dkey_count = 2
        akey_count = 1
        insert_objects(
            context=self.context, container=self.container, object_count=object_count ,
            dkey_count=dkey_count, akey_count=akey_count , base_dkey=self.random_dkey,
```

**Review by daltonbohning — COMMENTED** [review]
LGTM, except some minor comments

**Review by daltonbohning — APPROVED** [review]

File: src/tests/ftest/recovery/ddb.py (hunk 14 above)
Comment (shimizukko) [reply src/tests/ftest/recovery/ddb.py]: Fixed

File: src/tests/ftest/recovery/ddb.py [hunk 15]
```diff
@@ -0,0 +1,601 @@
...
+            if "vos" in file:
+                self.log.info("vos_file: %s", file)
+                return file
+
+        self.fail("vos file wasn't found in /mnt/daos/{}".format(self.pool.uuid.lower()))
+
+        return None  # to appease pylint
+
+    def test_recovery_ddb_ls(self):
+        """Test ddb ls.
```
Comment (dinghwah) [inline src/tests/ftest/recovery/ddb.py:57]: You might want to add JIRA DAOS-10830 to all the def test_.. description

**Review by dinghwah — APPROVED** [review]

File: src/tests/ftest/recovery/ddb.py (hunk 15 above)
Comment (shimizukko) [reply src/tests/ftest/recovery/ddb.py:57]: I'll update in the subsequent PR. Thanks.

**Pull request merged**